### ISSUE TITLE
`no-redeclare` not detecting redefinition of `for...of` variable

### ISSUE THREAD

**Issue by oliverwoodings** [eslint/eslint#7900]
**Tell us about your environment**

* **ESLint Version:** 3.8.1 and the online demo environment
* **Node Version:** 6.2
* **npm Version:** 3

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```js
{
  env: {
    es6: true
  },
  rules: {
    'no-redeclare': 2,
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
```js
  for (let foo of bar) {
    const foo = 'bar'
  }
```

**What did you expect to happen?**
I expected to receive an error about line to redeclaring `foo`

**What actually happened? Please include the actual, raw output from ESLint.**
I got no warnings or errors (screenshot is of the demo tool with only the `no-redeclare` rule enabled):

![screen shot 2017-01-11 at 10 32 34](https://cloud.githubusercontent.com/assets/621323/21845124/4c3c1ec8-d7e9-11e6-873b-a69ade6c84ee.png)

**Comment by not-an-aardvark** [comment]
I'm not sure about this, but I think the `foo` declared in the loop body is actually a different variable than the one declared in the loop header. So this would fall under [`no-shadow`](http://eslint.org/docs/rules/no-shadow) instead.

If `foo` was referring to the same variable in both cases, I would expect your code to throw a syntax error for the same reason as this:

```js
let foo = 1;
const foo = 1; // SyntaxError: Identifier 'foo' has already been declared
```

But in reality, I think the code is more similar to this:

```js
let foo = 1;
{
  const foo = 1; // creates a new `foo` variable that shadows the other one in this block
}
```

**Comment by oliverwoodings** [comment]
I think what is unclear is where the block scope starts. Buble throws a parsing exception for this, which is why I ended up looking into it. I guess this is maybe an issue with spec interpretation? Either Buble or Eslint have interpreted the spec incorrectly, but tbh I'm not sure which!

**Comment by not-an-aardvark** [comment]
Well, there's actually a [bug](https://github.com/ternjs/acorn/issues/487) in ESLint's parser where an error isn't even thrown for the first example. I was basing what I said off the fact that the code is valid syntax in Chrome, but I can look at the spec a bit later to double-check.

**Comment by oliverwoodings** [comment]
Ah thats a fair point - Chrome is probably a better indicator than Buble for spec compliance!

**Comment by not-an-aardvark** [comment]
Closing this issue because the `no-shadow` rule correctly reports this case.
